convert/changeset.go

```diff
@@ -0,0 +1,16 @@
+package convert
+
+// ChangeDetector defines funcs for getting a changeset from two
+// instances of a class. This interface has to be implemented by
+// all
+type ChangeDetector interface {
+	ChangeSet(older ChangeDetector) ([]Change, error)
+}
+
+// Change defines a set of changed values in an entity. It holds
+// the attribute name as the key and old and new values.
+type Change struct {
```

Comment:
Maybe this is also better modeled as an interface.

Reply:
You should really model this with an `interface`:

```go
type Change interface {
   AttributeName() string
   NewValue() interface{}
   OldValue() interface{}
}
```

Then you can create Generic changes or more complicated ones that even have values attached to them other than the old and new value.

Reply:
I don't think this makes sense. Changes are very very atomic. It intentionally simple. Plus, this would mean that we need to use getter and setter to read/write the values, which would make the already complex code on the rules even more complex and more hard to read.

Reply:
@michaelkleinhenz The code that produces a change is usually the domain code (e.g. iteration, label, area, work item, etc.) right?

That code wouldn't have to use getters or setters it can just create a change object of their own liking and then pass it to your system. Your system then receives it as a `Change` interface.

If I'm not mistaken your system treats the change as read-only, right? At least that's why I would expect. If that's true then things would become much clearer because not only is a change **atomic** then but it will also become **immutable**, which is great.

Reply:
I agree. The issue here is that the Change is something that should be used as part of a very "generic" API interface from very different areas of the code. In a full usage, this will be used from very different areas of the WIT code. So the goal needs to be to make this as robust as possible. on the other hand, Interfaces in golang are the opposite of robust as they can be carrying both literals and pointers, which has already brought a lot of fun to that implementation. With interfaces, type safety is compromised which is not a good idea given that this is used from many different people and areas in the WIT. 

I even thought about moving from an Interface to a struct for the `ChangeDetector` to make it more robust. For the `ChangeDetector` it makes even more sense to use a type instead of an Interface as this is the return type for the rules execution.

Reply:
I strongly disagree.

Reply:
That's unfortunate.

Reply:
@michaelkleinhenz I have to agree with @kwk . I think making this an interface makes much more sense in the long run. An interface makes this easily extensible and in a growing product this is incredibly important. I also don't see getters and setters are complex... getters and setters make the code more readable and easy to maintain.

Reply:
Let's have a conversation on golang interfaces. I think using interfaces here will make things more complex and more error-prone as an interface can hold a literal or a pointer - and that may change on runtime. If we use interfaces and need to be consistent with the actual instance, we need to add additional checks and complexity here. 

Yes, interfaces are extensible, but to be honest, that is not likely to change even in weird usecases. A similar definition in Angular (where I did take this) is stable for years now. Also, using interfaces here will make it less robust. And I would value robustness here higher than extensibility.

But I agree that this is a worthy discussion.